### PR TITLE
HDDS-7687. Support OM transfer Ratis leadership

### PR DESCRIPTION
**Pull request by Xushaohong**
## What changes were proposed in this pull request?

Support OM transfer Ratis leadership.
Split From https://issues.apache.org/jira/browse/HDDS-5686

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7687

## How was this patch tested?
UT and manual test.
image

### CODE REVIEW

**Comment by Xushaohong** [comment]
@szetszwo PTAL

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 1]
```diff
@@ -3023,6 +3025,51 @@ public ServiceInfoEx getServiceInfo() throws IOException {
...
+    try {
+      RaftGroupId groupID = omRatisServer.getRaftGroup().getGroupId();
+      RaftServer.Division division = omRatisServer.getServer()
+          .getDivision(groupID);
+      RaftPeerId targetPeerId;
+      if (nodeId.isEmpty()) {
+        RaftPeer curLeader = omRatisServer.getLeader();
+        targetPeerId = division.getGroup()
+            .getPeers().stream().filter(a -> !a.equals(curLeader)).findFirst()
+            .map(RaftPeer::getId).orElse(null);
```
Comment (szetszwo) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:3052]: Use `orElseThrow`.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/TransferOmLeaderSubCommand.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      description = "Ozone Manager Service ID."
+  )
+  private String omServiceId;
+
+  @CommandLine.ArgGroup(multiplicity = "1")
+  private TransferOption configGroup;
+
+  static class TransferOption {
+    @CommandLine.Option(
+        names = {"-n", "--nodeId"},
```
Comment (szetszwo) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/TransferOmLeaderSubCommand.java:51]: Use `transfereeId` or `newLeaderId`.

BTW, is there a way to specify the which OM to send the command?

File: hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot [hunk 3]
```diff
@@ -182,4 +194,38 @@ Restart OM and Verify Ratis Logs
...
+    # Find Leader OM and one Follower OM
+    ${leaderOM} =           Get OM Leader Node
+                            LOG                     Leader OM: ${leaderOM}
+    # Transfer leadership to the Follower OM
+    ${result} =             Execute                 ozone admin om transfer -r
+                            LOG                     ${result}
+                            Should Contain          ${result}               Transfer leadership successfully
+
+    ${newLeaderOM} =        Get OM Leader Node
+                            Should Not be Equal     ${leaderOM}             ${newLeaderOM}
```
Comment (szetszwo) [inline hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot:230]: Let add a test case with a specific new leader.  Just transfer it back to `${leaderOM}`.

**Review by szetszwo — COMMENTED** [review]
@Xushaohong , thanks a lot for separating this out from HDDS-5686!

- `nodeId` may be confusing -- it may refer to the server receiving the command.  Please replace it with `transfereeId` or `newLeaderId`.

See also the comments inlined.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java [hunk 4]
```diff
@@ -488,4 +490,62 @@ public static void debug(ByteBuf buf, String name, Logger log) {
...
+   * Use raft client to send admin request, transfer the leadership.
+   * 1. Set priority and send setConfiguration request
+   * 2. Trigger transferLeadership API.
+   *
+   * @param division      the Raft division
+   * @param targetPeerId  the target expected leader
+   * @throws IOException
+   */
+  public static void transferRatisLeadership(ConfigurationSource conf,
+      RaftServer.Division division, RaftPeerId targetPeerId)
```
[the remark this answers is not in the document]
Comment (Xushaohong) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java]: done

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java [hunk 5]
```diff
@@ -488,4 +490,62 @@ public static void debug(ByteBuf buf, String name, Logger log) {
...
+      throws IOException {
+    RaftGroup raftGroup = division.getGroup();
+    // TODO: need a common raft client related conf.
+    RaftClient raftClient = newRaftClient(SupportedRpcType.GRPC, null,
```
[the remark this answers is not in the document]
Comment (Xushaohong) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/ratis/RatisHelper.java]: updated

File: hadoop-hdds/interface-client/src/main/proto/hdds.proto [hunk 6]
```diff
@@ -452,3 +452,10 @@ message ContainerBalancerConfigurationProto {
     required bool shouldRun = 18;
     optional int32 nextIterationIndex = 19;
 }
+
+message TransferLeadershipRequestProto {
+    required string nodeId = 1;
```
[the remark this answers is not in the document]
Comment (Xushaohong) [reply hadoop-hdds/interface-client/src/main/proto/hdds.proto]: Got it

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java (hunk 1 above)
Comment (Xushaohong) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java]: done

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/TransferOmLeaderSubCommand.java (hunk 2 above)
Comment (Xushaohong) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/TransferOmLeaderSubCommand.java]: > BTW, is there a way to specify the which OM to send the command?

This is the client's function. Currently, it seems not supported.

File: hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot (hunk 3 above)
Comment (Xushaohong) [reply hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot:230]: > Let add a test case with a specific new leader. Just transfer it back to `${leaderOM}`.

In robot test, there is already a test case to transfer to the specific leader and also a random cast to transfer randomly.

https://github.com/apache/ozone/blob/c763ee4e8296ca896689813771745758e55aea2e/hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot#L197

Comment (szetszwo) [reply hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot:230]: I see.  Thanks.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good

File: hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot (hunk 3 above)
Comment (Xushaohong) [reply hadoop-ozone/dist/src/main/smoketest/omha/testOMHA.robot:230]: > `testOMHA.robot` is not run in CI, as it is only part of the `ozone-om-ha` environment, which is disabled:

@adoroszlai I see, the robot test could be left to be fixed.

**Pull request merged**